subprojects/logging/src/main/java/org/gradle/internal/featurelifecycle/LoggingDeprecatedFeatureHandler.java

```diff
@@ -45,36 +47,39 @@ public class LoggingDeprecatedFeatureHandler implements FeatureHandler<Deprecate
 
     private final Set<String> messages = new HashSet<String>();
     private boolean deprecationsFound = false;
-    private UsageLocationReporter locationReporter;
+    private ProblemDiagnosticsFactory problemDiagnosticsFactory = new NoOpProblemDiagnosticsFactory();
 
     private WarningMode warningMode = WarningMode.Summary;
     private BuildOperationProgressEventEmitter progressEventEmitter;
     private GradleException error;
 
-    public LoggingDeprecatedFeatureHandler() {
-        this.locationReporter = DoNothingReporter.INSTANCE;
-    }
-
-    public void init(UsageLocationReporter reporter, WarningMode warningMode, BuildOperationProgressEventEmitter progressEventEmitter) {
-        this.locationReporter = reporter;
+    public void init(ProblemDiagnosticsFactory problemDiagnosticsFactory, WarningMode warningMode, BuildOperationProgressEventEmitter progressEventEmitter) {
+        this.problemDiagnosticsFactory = problemDiagnosticsFactory;
         this.warningMode = warningMode;
         this.progressEventEmitter = progressEventEmitter;
     }
 
     @Override
-    public void featureUsed(DeprecatedFeatureUsage usage) {
+    public void featureUsed(final DeprecatedFeatureUsage usage) {
         deprecationsFound = true;
+        ProblemDiagnostics diagnostics = problemDiagnosticsFactory.forCurrentCaller(new ProblemDiagnosticsFactory.StackTraceTransformer() {
+            @Override
+            public List<StackTraceElement> transform(StackTraceElement[] original) {
+                return FeatureUsage.calculateStack(usage.getCalledFrom(), original);
+            }
+        });
         if (warningMode.shouldDisplayMessages()) {
             String featureMessage = usage.formattedMessage();
             StringBuilder message = new StringBuilder();
-            locationReporter.reportLocation(usage, message);
-            if (message.length() > 0) {
+            Location location = diagnostics.getLocation();
+            if (location != null) {
+                message.append(location.getFormatted());
                 message.append(SystemProperties.getInstance().getLineSeparator());
             }
             message.append(featureMessage);
-            displayDeprecationIfSameMessageNotDisplayedBefore(message, usage.getStack());
+            displayDeprecationIfSameMessageNotDisplayedBefore(message, diagnostics.getStack());
```

Comment:
:question: May this cause the message to be displayed an additional time as before, when we have a lot of deprecations with the same stack trace? We'd show it once with the (possible truncated) stacktrace and once without, right? I suppose that isn't a big deal.

Reply:
Good point. I'll fix that.

Reply:
Done